### ISSUE TITLE
`-Wunreachable-code` and coroutines.

### ISSUE THREAD

**Issue by asoffer** [llvm/llvm-project#69219]
The `-Wunreachable-code` warning triggers on code of this form:

```
task f() {
  std::abort();
  co_return;
}
```

While it's true that `co_return` is never executed, one cannot remove the `co_return` without changing the semantics of the program. This is the simplest mechanism one can use to write a never-returning coroutine, and should therefore not be flagged. As a current workaround, one can spell this as 
```
task f() {
    co_return [&] { std::abort(); }();
}
```
but ideally when unreachable code is used to determine the semantics of the function/coroutine, it should not trigger `-Wunreachable-code` warning.

Reproduction:
https://godbolt.org/z/7EoEcf4M9

**Comment by shafik** [comment]
CC @bcardosolopes

**Comment by jacobsa** [comment]
Thanks for filing, Andy. This is synthesized from a Google-internal bug report I wrote, so I'll try to describe the motivation for the example in more detail.

This is from a test, where I have a coroutine that I want to assert is _not_ resumed from its initial suspend point. I want the test to crash if it is resumed. The example in the initial bug report will do so, assuming the promise suspends there:

```c++
// Assume `task::promise_type` suspends in initial_suspend.
// Then it's okay to call this, just not resume it.
task f() {
  std::abort();
  co_return;
}
```

That can be called without crashing, just not resumed afterward. But if you remove the "unused" code, then you get a function that doesn't have an initial suspend point, just crashing as soon as it's called:

```c++
// Not a coroutine; will crash immediately when called.
task f() {
  std::abort();
}
```

**Comment by jacobsa** [comment]
In other words, the warning is telling me to change the semantics of my program. I don't think it should do this—it should only warn for code that can be removed without changing semantics.

**Comment by yuxuanchen1997** [comment]
Can you write 
```
task f() {
  co_await std::suspend_never{};
  std::abort();
}
```

Instead?

**Comment by jacobsa** [comment]
In this case, no. The promise type doesn't support awaiting arbitrary coroutine handles, only other handles of its type.

Regardless, that would only be a workaround; the warning is still incorrectly complaining that the code is unused (in the sense that it doesn't change semantics to remove it).

**Comment by hokein** [comment]
I think we can probably fix it by suppressing the unreachable warning if the source location of the unreachable block is before the first coroutine statement of the current function, we can add the filtering in https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/AnalysisBasedWarnings.cpp#L84C1-L84C1.

**Comment by dwblaikie** [comment]
> I think we can probably fix it by suppressing the unreachable warning if the source location of the unreachable block is before the first coroutine statement of the current function, we can add the filtering in https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/AnalysisBasedWarnings.cpp#L84C1-L84C1.

I don't think source locations would be adequate - macros can often/easily mess that sort of thing up/make it hard to determine what comes first.

A more structured/AST-based approach is probably required. (if I had to guess where to start, I'd probably look at the maybe/sometimes uninitialized warnings to see how they understand the relationship between different statements to make some similar determination here)
